compiler/rustc_typeck/src/check/wfcheck.rs

```diff
@@ -201,6 +201,40 @@ pub fn check_trait_item(tcx: TyCtxt<'_>, def_id: LocalDefId) {
     };
     check_object_unsafe_self_trait_by_name(tcx, &trait_item);
     check_associated_item(tcx, trait_item.hir_id(), span, method_sig);
+
+    let encl_trait_hir_id = tcx.hir().get_parent_item(hir_id);
+    let encl_trait = tcx.hir().expect_item(encl_trait_hir_id);
+    if [tcx.lang_items().fn_trait(), tcx.lang_items().fn_mut_trait()]
+        .contains(&Some(encl_trait.def_id.to_def_id()))
```

Comment:
Could we move this check out so that we can identify which of the two traits was found? That way we can then make the description completely accurate ("first argument of `call` in `fn` lang item must be a reference").

Reply:
Done, along with your other suggestions. Thanks for reviewing this!